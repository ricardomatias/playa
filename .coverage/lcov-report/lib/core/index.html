<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/core</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/core
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.17% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>343/389</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.98% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>141/172</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>57/68</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.83% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>332/378</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Chord.js"><a href="Chord.js.html">Chord.js</a></td>
	<td data-value="96.9" class="pic high"><div class="chart"><div class="cover-fill" style="width: 96%;"></div><div class="cover-empty" style="width:4%;"></div></div></td>
	<td data-value="96.9" class="pct high">96.9%</td>
	<td data-value="129" class="abs high">125/129</td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="44" class="abs high">40/44</td>
	<td data-value="94.44" class="pct high">94.44%</td>
	<td data-value="18" class="abs high">17/18</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="124" class="abs high">120/124</td>
	</tr>

<tr>
	<td class="file high" data-value="HarmonyBase.js"><a href="HarmonyBase.js.html">HarmonyBase.js</a></td>
	<td data-value="84.62" class="pic high"><div class="chart"><div class="cover-fill" style="width: 84%;"></div><div class="cover-empty" style="width:16%;"></div></div></td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	</tr>

<tr>
	<td class="file high" data-value="Key.js"><a href="Key.js.html">Key.js</a></td>
	<td data-value="94.67" class="pic high"><div class="chart"><div class="cover-fill" style="width: 94%;"></div><div class="cover-empty" style="width:6%;"></div></div></td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="36" class="abs high">30/36</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	</tr>

<tr>
	<td class="file low" data-value="Mode.js"><a href="Mode.js.html">Mode.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	</tr>

<tr>
	<td class="file high" data-value="Note.js"><a href="Note.js.html">Note.js</a></td>
	<td data-value="89.74" class="pic high"><div class="chart"><div class="cover-fill" style="width: 89%;"></div><div class="cover-empty" style="width:11%;"></div></div></td>
	<td data-value="89.74" class="pct high">89.74%</td>
	<td data-value="78" class="abs high">70/78</td>
	<td data-value="90.48" class="pct high">90.48%</td>
	<td data-value="42" class="abs high">38/42</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	<td data-value="89.61" class="pct high">89.61%</td>
	<td data-value="77" class="abs high">69/77</td>
	</tr>

<tr>
	<td class="file high" data-value="Scale.js"><a href="Scale.js.html">Scale.js</a></td>
	<td data-value="93.94" class="pic high"><div class="chart"><div class="cover-fill" style="width: 93%;"></div><div class="cover-empty" style="width:7%;"></div></div></td>
	<td data-value="93.94" class="pct high">93.94%</td>
	<td data-value="66" class="abs high">62/66</td>
	<td data-value="82.86" class="pct high">82.86%</td>
	<td data-value="35" class="abs high">29/35</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="93.44" class="pct high">93.44%</td>
	<td data-value="61" class="abs high">57/61</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="types.js"><a href="types.js.html">types.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 15:09:45 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
