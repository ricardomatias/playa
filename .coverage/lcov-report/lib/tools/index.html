<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/tools</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/tools
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">91.08% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>296/325</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">85% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>102/120</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">84.75% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>50/59</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">90.76% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>275/303</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="choose.js"><a href="choose.js.html">choose.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

<tr>
	<td class="file high" data-value="distance.js"><a href="distance.js.html">distance.js</a></td>
	<td data-value="97.37" class="pic high"><div class="chart"><div class="cover-fill" style="width: 97%;"></div><div class="cover-empty" style="width:3%;"></div></div></td>
	<td data-value="97.37" class="pct high">97.37%</td>
	<td data-value="38" class="abs high">37/38</td>
	<td data-value="94.44" class="pct high">94.44%</td>
	<td data-value="18" class="abs high">17/18</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="97.14" class="pct high">97.14%</td>
	<td data-value="35" class="abs high">34/35</td>
	</tr>

<tr>
	<td class="file high" data-value="distribute.js"><a href="distribute.js.html">distribute.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="24" class="abs high">24/24</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="20" class="abs high">20/20</td>
	</tr>

<tr>
	<td class="file high" data-value="euclidean.js"><a href="euclidean.js.html">euclidean.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="20" class="abs high">20/20</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="18" class="abs high">18/18</td>
	</tr>

<tr>
	<td class="file high" data-value="friendly.js"><a href="friendly.js.html">friendly.js</a></td>
	<td data-value="97.56" class="pic high"><div class="chart"><div class="cover-fill" style="width: 97%;"></div><div class="cover-empty" style="width:3%;"></div></div></td>
	<td data-value="97.56" class="pct high">97.56%</td>
	<td data-value="82" class="abs high">80/82</td>
	<td data-value="92.59" class="pct high">92.59%</td>
	<td data-value="27" class="abs high">25/27</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="13" class="abs high">13/13</td>
	<td data-value="97.4" class="pct high">97.4%</td>
	<td data-value="77" class="abs high">75/77</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="interval.js"><a href="interval.js.html">interval.js</a></td>
	<td data-value="88.89" class="pic high"><div class="chart"><div class="cover-fill" style="width: 88%;"></div><div class="cover-empty" style="width:12%;"></div></div></td>
	<td data-value="88.89" class="pct high">88.89%</td>
	<td data-value="36" class="abs high">32/36</td>
	<td data-value="85" class="pct high">85%</td>
	<td data-value="20" class="abs high">17/20</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="91.43" class="pct high">91.43%</td>
	<td data-value="35" class="abs high">32/35</td>
	</tr>

<tr>
	<td class="file high" data-value="random.js"><a href="random.js.html">random.js</a></td>
	<td data-value="93.33" class="pic high"><div class="chart"><div class="cover-fill" style="width: 93%;"></div><div class="cover-empty" style="width:7%;"></div></div></td>
	<td data-value="93.33" class="pct high">93.33%</td>
	<td data-value="15" class="abs high">14/15</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="92.86" class="pct high">92.86%</td>
	<td data-value="14" class="abs high">13/14</td>
	</tr>

<tr>
	<td class="file medium" data-value="ring.js"><a href="ring.js.html">ring.js</a></td>
	<td data-value="77.27" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 77%;"></div><div class="cover-empty" style="width:23%;"></div></div></td>
	<td data-value="77.27" class="pct medium">77.27%</td>
	<td data-value="22" class="abs medium">17/22</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="5" class="abs low">2/5</td>
	<td data-value="77.27" class="pct medium">77.27%</td>
	<td data-value="22" class="abs medium">17/22</td>
	</tr>

<tr>
	<td class="file high" data-value="roll.js"><a href="roll.js.html">roll.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="rotate.js"><a href="rotate.js.html">rotate.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	</tr>

<tr>
	<td class="file high" data-value="shuffle.js"><a href="shuffle.js.html">shuffle.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="12" class="abs high">12/12</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file medium" data-value="time.js"><a href="time.js.html">time.js</a></td>
	<td data-value="71.93" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 71%;"></div><div class="cover-empty" style="width:29%;"></div></div></td>
	<td data-value="71.93" class="pct medium">71.93%</td>
	<td data-value="57" class="abs medium">41/57</td>
	<td data-value="67.86" class="pct medium">67.86%</td>
	<td data-value="28" class="abs medium">19/28</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="10" class="abs medium">5/10</td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="56" class="abs medium">40/56</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 15:09:45 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
