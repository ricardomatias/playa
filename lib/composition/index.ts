export { createMotif } from './motif';
export { createChordProgression } from './chord-progression';
export { createMelodies } from './melodies';
export { createPercussion } from './percussion';
export { createArp, genHarmonicShift } from './arp';

export * as Rhythm from './rhythm';
export * from './sequence';
export * from './movement';
