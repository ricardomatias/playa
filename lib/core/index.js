/**
* @namespace {module} Core
*/
export { Note } from './Note';
export { default as Chord } from './Chord';
export { default as Scale } from './Scale';
export { default as Key } from './Key';
export { Time } from './Time';
export { Phrase } from './Phrase';
export { PianoRoll } from './PianoRoll';

