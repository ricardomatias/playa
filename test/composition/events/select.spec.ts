import { describe, it, expect } from 'vitest';
import { Rhythm, Events, createMotif, createArp } from '../../../lib/composition';
import { Scale } from '../../../lib/core';
import { convertEventsToNotevalues, random } from '../../../lib/tools';
import '../../matchers';

describe('#select', () => {
	it('should select every 2n before after 4n', () => {
		// given
		random.setSeed('test');

		const rhythm = Rhythm.free('4:0:0', ['4n']);

		// when
		const after = Events.select(rhythm).every('2n').after('4n').rest();
		const before = Events.select(rhythm).before('2m').rest();

		expect(after).toMatchSnapshot();
		expect(before).toMatchSnapshot();
	});

	it('should select before after 2n', () => {
		// given
		random.setSeed('test');

		const rhythm = Rhythm.free('1:0:0', ['8n']);

		// when
		const after = Events.select(rhythm).after('2n').rest();
		const before = Events.select(rhythm).before('2n').rest();

		expect(after).toMatchSnapshot();
		expect(before).toMatchSnapshot();
	});

	it('should select applyRest', () => {
		// given
		random.setSeed('test');

		const rhythm = Rhythm.free('1:0:0', ['8n']);

		// when
		const results = Events.select(rhythm)
			.before('2n')
			.applyRest(() => random.boolean());

		expect(results).toMatchSnapshot();
	});

	it('should select multiple', () => {
		// given
		random.setSeed('test');

		const rhythm = Rhythm.free('1:0:0', ['8n']);

		// when
		const results = Events.select(rhythm).before('0:2:0').after('0:1:0').rest();
		const results2 = Events.select(rhythm).between('0:1:0', '0:2:0').rest();
		const results3 = Events.select(Rhythm.free('4:0:0', ['4n']))
			.every('1:0:0')
			.between('0:1:0', '0:2:0')
			.rest();

		expect(results).toMatchObject(results2);
		expect(results).toMatchSnapshot();
		expect(results3).toMatchSnapshot();
	});

	it('should insert simple', () => {
		// given
		random.setSeed('test');

		const rhythm = Rhythm.free('2:0:0', ['8n']);
		const rhythm2 = Rhythm.free('2n', ['4n']);

		const notes = new Scale('A', Scale.Major).notes;

		const motif = createMotif(notes, rhythm);
		const motif2 = createMotif(notes, rhythm2);

		// when
		const results = Events.select(motif).every('1:0:0').after('2n').insert(motif2);

		expect(convertEventsToNotevalues(results)).toEqual([
			'8n',
			'8n',
			'8n',
			'8n',
			'4n',
			'4n',
			'8n',
			'8n',
			'8n',
			'8n',
			'4n',
			'4n',
		]);
	});

	it('should insert complex', () => {
		// given
		random.setSeed('test');

		const scale = new Scale('A', Scale.Major);

		const motif = Events.expand(createArp(scale, [1, 3, '-7'], ['4n', '4nd']), '2m');
		const rhythm2 = Rhythm.free('2n', Rhythm.Presets.Mixed);
		const motif2 = createMotif(scale.notes, rhythm2);

		// when
		const results = Events.select(motif).before('1m').after('2n').insert(motif2);

		expect(convertEventsToNotevalues(motif)).toMatchInlineSnapshot(`
		[
		  "4n",
		  "4nd",
		  "4n",
		  "4n",
		  "4nd",
		  "4n",
		  "4n",
		]
	`);
		expect(convertEventsToNotevalues(results)).toMatchInlineSnapshot(`
		[
		  "4n",
		  "4nd",
		  "4nd",
		  "16n",
		  "16n",
		  "4nd",
		  "4n",
		  "4n",
		]
	`);
	});
});
