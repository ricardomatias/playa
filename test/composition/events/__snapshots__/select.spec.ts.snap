// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#select > should select applyRest 1`] = `
[
  {
    "dur": 240,
    "isRest": true,
    "next": 240,
    "time": 0,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 480,
    "time": 240,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 720,
    "time": 480,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 960,
    "time": 720,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1200,
    "time": 960,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1440,
    "time": 1200,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1680,
    "time": 1440,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1920,
    "time": 1680,
  },
]
`;

exports[`#select > should select before after 2n 1`] = `
[
  {
    "dur": 240,
    "isRest": false,
    "next": 240,
    "time": 0,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 480,
    "time": 240,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 720,
    "time": 480,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 960,
    "time": 720,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 1200,
    "time": 960,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 1440,
    "time": 1200,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 1680,
    "time": 1440,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 1920,
    "time": 1680,
  },
]
`;

exports[`#select > should select before after 2n 2`] = `
[
  {
    "dur": 240,
    "isRest": true,
    "next": 240,
    "time": 0,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 480,
    "time": 240,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 720,
    "time": 480,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 960,
    "time": 720,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1200,
    "time": 960,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1440,
    "time": 1200,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1680,
    "time": 1440,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1920,
    "time": 1680,
  },
]
`;

exports[`#select > should select every 2n before after 4n 1`] = `
[
  {
    "dur": 480,
    "isRest": false,
    "next": 480,
    "time": 0,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 960,
    "time": 480,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 1440,
    "time": 960,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 1920,
    "time": 1440,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 2400,
    "time": 1920,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 2880,
    "time": 2400,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 3360,
    "time": 2880,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 3840,
    "time": 3360,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 4320,
    "time": 3840,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 4800,
    "time": 4320,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 5280,
    "time": 4800,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 5760,
    "time": 5280,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 6240,
    "time": 5760,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 6720,
    "time": 6240,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 7200,
    "time": 6720,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 7680,
    "time": 7200,
  },
]
`;

exports[`#select > should select every 2n before after 4n 2`] = `
[
  {
    "dur": 480,
    "isRest": true,
    "next": 480,
    "time": 0,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 960,
    "time": 480,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 1440,
    "time": 960,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 1920,
    "time": 1440,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 2400,
    "time": 1920,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 2880,
    "time": 2400,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 3360,
    "time": 2880,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 3840,
    "time": 3360,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 4320,
    "time": 3840,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 4800,
    "time": 4320,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 5280,
    "time": 4800,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 5760,
    "time": 5280,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 6240,
    "time": 5760,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 6720,
    "time": 6240,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 7200,
    "time": 6720,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 7680,
    "time": 7200,
  },
]
`;

exports[`#select > should select multiple 1`] = `
[
  {
    "dur": 240,
    "isRest": false,
    "next": 240,
    "time": 0,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 480,
    "time": 240,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 720,
    "time": 480,
  },
  {
    "dur": 240,
    "isRest": true,
    "next": 960,
    "time": 720,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1200,
    "time": 960,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1440,
    "time": 1200,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1680,
    "time": 1440,
  },
  {
    "dur": 240,
    "isRest": false,
    "next": 1920,
    "time": 1680,
  },
]
`;

exports[`#select > should select multiple 2`] = `
[
  {
    "dur": 480,
    "isRest": false,
    "next": 480,
    "time": 0,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 960,
    "time": 480,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 1440,
    "time": 960,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 1920,
    "time": 1440,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 2400,
    "time": 1920,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 2880,
    "time": 2400,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 3360,
    "time": 2880,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 3840,
    "time": 3360,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 4320,
    "time": 3840,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 4800,
    "time": 4320,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 5280,
    "time": 4800,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 5760,
    "time": 5280,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 6240,
    "time": 5760,
  },
  {
    "dur": 480,
    "isRest": true,
    "next": 6720,
    "time": 6240,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 7200,
    "time": 6720,
  },
  {
    "dur": 480,
    "isRest": false,
    "next": 7680,
    "time": 7200,
  },
]
`;
